server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1587,6 +1590,11 @@ static Engine.Searcher wrapSearcher(
         ); // completes stats recording
     }
 
+    public void setGlobalCheckpointIfUnpromotable(long globalCheckpoint) {
+        assert shardRouting.isPromotableToPrimary() == false : "must only call this on unpromotable shards";
+        globalCheckPointIfUnpromotable = globalCheckpoint;
```

Comment:
I wonder if we should also assert something like `DiscoveryNode.isStateless(indexShard.indexSettings().getNodeSettings())` here, just to be paranoid?

Reply:
I'm confident enough in just asserting `shardRouting.isPromotableToPrimary() == false`. All manner of other things won't work if we somehow end up with an unpromotable shard in a regular ES cluster.